plugin/pkg/scheduler/algorithm/predicates/predicates_test.go

```diff
@@ -557,7 +557,180 @@ func TestPodFitsSelector(t *testing.T) {
+									{Key: "key2", Operator: api.NodeSelectorOpNotIn, Values: []string{"value1", "value2"}},
+								},
+							},
+						},
+					},
+				},
+			},
+			labels: map[string]string{
+				"foo": "bar",
+			},
+			fits: true,
+			test: "same labels with proper Affinity",
+		},
+		{
+			pod: &api.Pod{
+				Spec: api.PodSpec{
+					NodeSelector: map[string]string{
+						"foo": "bar",
+					},
+					Affinity: &api.Affinity{
+						HardNodeAffinity: &api.NodeSelector{
+							NodeSelectorTerms: []api.NodeSelectorTerm{
+								{
+									MatchExpressions: []api.NodeSelectorRequirement{
+										{Key: "foo", Operator: api.NodeSelectorOpIn, Values: []string{"value1", "value2"}},
+									},
+								},
+							},
+						},
+						SoftNodeAffinity: []api.SoftNodeAffinityTerm{
+							{
+								Weight: 2,
+								MatchExpressions: []api.NodeSelectorRequirement{
+									{Key: "key2", Operator: api.NodeSelectorOpNotIn, Values: []string{"value1", "value2"}},
+								},
+							},
+						},
+					},
+				},
+			},
+			labels: map[string]string{
+				"foo": "bar",
+			},
+			fits: false,
+			test: "node labels don't match pod hard node Affinity requirements",
+		},
+		{
+			pod: &api.Pod{
+				Spec: api.PodSpec{
+					NodeSelector: map[string]string{
+						"foo": "bar",
+					},
+					Affinity: &api.Affinity{
+						HardNodeAffinity: &api.NodeSelector{
+							NodeSelectorTerms: []api.NodeSelectorTerm{
+								{
+									MatchExpressions: []api.NodeSelectorRequirement{
+										{Key: "foo", Operator: api.NodeSelectorOpIn, Values: []string{"bar", "value2"}},
+									},
+								},
+								{
+									MatchExpressions: []api.NodeSelectorRequirement{
+										{Key: "diffkey", Operator: api.NodeSelectorOpIn, Values: []string{"wrong", "value2"}},
+									},
+								},
+							},
+						},
+						SoftNodeAffinity: []api.SoftNodeAffinityTerm{
+							{
+								Weight: 2,
+								MatchExpressions: []api.NodeSelectorRequirement{
+									{Key: "key2", Operator: api.NodeSelectorOpNotIn, Values: []string{"value1", "value2"}},
+								},
+							},
+						},
+					},
+				},
+			},
+			labels: map[string]string{
+				"foo": "bar",
+			},
+			fits: true,
+			test: "same labels with Affinity with multiple NodeSelectTerms",
+		},
+		{
+			pod: &api.Pod{
+				Spec: api.PodSpec{
+					NodeSelector: map[string]string{
+						"foo": "bar",
+					},
+					Affinity: &api.Affinity{
+						HardNodeAffinity: &api.NodeSelector{},
+					},
+				},
+			},
+			labels: map[string]string{
+				"foo": "bar",
+			},
+			fits: false,
+			test: "same labels with Affinity with nil NodeSelectTerms",
+		},
+		{
+			pod: &api.Pod{
+				Spec: api.PodSpec{
+					NodeSelector: map[string]string{
+						"foo": "bar",
+					},
+					Affinity: &api.Affinity{
+						HardNodeAffinity: &api.NodeSelector{NodeSelectorTerms: []api.NodeSelectorTerm{}},
+					},
+				},
+			},
+			labels: map[string]string{
+				"foo": "bar",
+			},
+			fits: false,
```

Comment:
This doesn't sound right. Because
```
// An empty node selector term matches all objects. A null node selector term
// matches no objects.
type NodeSelectorTerm struct {
```

@davidopp This is serious problem..